src/NodeAnalyzer/ModelAnalyzer.php

```diff
@@ -52,26 +54,89 @@ public function getPrimaryKey(string $class): ?string
     }
 
     /**
+     * @param string|ObjectType $model
+     * @param string $methodName
+     * @return bool
+     */
+    public function isQueryScopeOnModel(string|ObjectType $model, string $scopeName, Scope $scope): bool
+    {
+        if (! is_string($model)) {
+            $model = $model->getClassName();
+        }
+
+        $classReflection = $this->getClass($model);
+
+        if ($classReflection->hasMethod('scope' . ucfirst($scopeName))) {
+            return true;
+        }
+
+        try {
+            $method = $classReflection->getMethod($scopeName, $scope);
+
+            if ($this->usesScopeAttribute($method))  {
+                return true;
+            }
+        } catch (MissingMethodFromReflectionException) {
+        }
+
+        return false;
+    }
+
+    private function usesScopeAttribute(ExtendedMethodReflection $methodReflection): bool
+    {
+        foreach ($methodReflection->getAttributes() as $attribute) {
+            if ($attribute->getName() === 'Illuminate\Database\Eloquent\Attributes\Scope') {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Get the ClassReflectionFor the Model
+     *
      * @param  class-string<Model>  $class
      *
-     * @throws Exception
+     * @throws InvalidArgumentException
      */
     private function getClass(string $class): ClassReflection
     {
         if (! $this->reflectionProvider->hasClass($class)) {
-            throw new Exception('Class not found');
+            throw new InvalidArgumentException('Class not found');
         }
 
         $classReflection = $this->reflectionProvider->getClass($class);
 
         if (! $classReflection->isClass()) {
-            throw new Exception('Class is not class');
+            throw new InvalidArgumentException('Class string does not resolve to class');
         }
 
         if (! $classReflection->isSubclassOfClass($this->reflectionProvider->getClass(Model::class))) {
-            throw new Exception('Class is not subclass of Model');
+            throw new InvalidArgumentException('Class is not subclass of Model');
         }
 
         return $classReflection;
     }
+
+    /**
+     * Create an instance of the Model to interact with
+     *
+     * @param class-string<Model>|ObjectType $model
+     * @return Model
+     * @throws ReflectionException
+     */
+    private function resolveModelClassToInstance(string|ObjectType $model): Model
+    {
+        if (! is_string($model)) {
+            $model = $model->getClassName();
+        }
+
+        $classReflection = $this->getClass($model);
```

Comment:
There's no need to get the class name from the object type and then get the class reflection when you can directly get the class reflection:


```suggestion
        $classReflection = is_string($model)
             ? $this->getClass($model)
             : $model->getObjectClassReflections()[0];
```